homeassistant/components/esphome/manager.py

```diff
@@ -1150,13 +1162,60 @@ class ServiceMetadata(NamedTuple):
 }
 
 
-@callback
-def execute_service(
-    entry_data: RuntimeEntryData, service: UserService, call: ServiceCall
-) -> None:
-    """Execute a service on a node."""
+async def execute_service(
+    entry_data: RuntimeEntryData,
+    service: UserService,
+    call: ServiceCall,
+    *,
+    supports_response: SupportsResponseType,
+) -> ServiceResponse:
+    """Execute a service on a node and optionally wait for response."""
+    # Determine if we should wait for a response
+    # NONE: fire and forget
+    # OPTIONAL/ONLY/STATUS: always wait for success/error confirmation
+    wait_for_response = supports_response != SupportsResponseType.NONE
+
+    if not wait_for_response:
+        # Fire and forget - no response expected
+        try:
+            entry_data.client.execute_service(service, call.data)
+        except APIConnectionError as err:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="action_call_failed",
+                translation_placeholders={
+                    "call_name": service.name,
+                    "device_name": entry_data.name,
+                    "error": str(err),
+                },
+            ) from err
+        return None
+
+    # Wait for response
+    call_id = next(_CALL_ID_COUNTER)
+    loop = asyncio.get_running_loop()
+    future: asyncio.Future[ExecuteServiceResponse] = loop.create_future()
+
+    def on_response(response: ExecuteServiceResponse) -> None:
+        if not future.done():
+            loop.call_soon_threadsafe(future.set_result, response)
```

Comment:
There shouldn't be a thread safety concern here since this is all running in the same thread. future.set_result(response) should be safe to call without `loop.call_soon_threadsafe`